[ARMv7] Avoid unaligned strd

justinmichaud opened the pull request:
#### 5e2002044e76e8a593ca5630ee6e71c760a01463
<pre>
[ARMv7] Avoid unaligned strd
<a href="https://bugs.webkit.org/show_bug.cgi?id=293694">https://bugs.webkit.org/show_bug.cgi?id=293694</a>

Reviewed by Yusuke Suzuki and Sosuke Suzuki.

Not all armv7 chips are required to suport unaligned strd instructions.
For example, my Neoverse N1 does not support it.

This is documented at <a href="https://developer.arm.com/documentation/100748/0624/Alignment-support-in-Arm-Compiler-for-Embedded-6/Aligned-and-unaligned-accesses.">https://developer.arm.com/documentation/100748/0624/Alignment-support-in-Arm-Compiler-for-Embedded-6/Aligned-and-unaligned-accesses.</a>

We should remove these lowerings to support all armv7 processors.
In the future, we may want to add them back when we can handle
the alignment check separately.

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::storePair32):

Canonical link: <a href="https://commits.webkit.org/298097@main">https://commits.webkit.org/298097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80bc265e9635c66e3767fa0a4e683b42c2ed364b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113201 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119409 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64538 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41499 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86163 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41537 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116148 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26828 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101836 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66483 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19962 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63163 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105718 "Built successfully and passed tests") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96213 "Found 1 new API test failure: TestWebKitAPI.WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithPersistedDomain (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122626 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111817 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30052 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95013 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98048 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94755 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24318 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39902 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17712 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36415 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40165 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45664 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136047 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39806 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36508 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43139 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41543 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a67ff9f1a7ccb535366770d33d57635b86ad6c13. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46003

aoikonomopoulos commented:
Is this on linux? AFAIU the linux kernel is supposed to cover up for unaligned `strd`/`ldrd` (see `do_alignment_ldrdstrd` in https://github.com/torvalds/linux/blob/master/arch/arm/mm/alignment.c (for a 32-bit kernel) and https://github.com/torvalds/linux/blob/master/arch/arm64/kernel/compat_alignment.c (for a 64-bit one)).

Still better not to have to take the fault of course, so I agree that this patch is the best way forward RN.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1fc26bd4092762e6b271c967ec061733ff368c27. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46003

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/80bc265e9635c66e3767fa0a4e683b42c2ed364b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46003

justinmichaud committed:
[ARMv7] Avoid unaligned strd
https://bugs.webkit.org/show_bug.cgi?id=293694

Reviewed by Yusuke Suzuki and Sosuke Suzuki.

Not all armv7 chips are required to suport unaligned strd instructions.
For example, my Neoverse N1 does not support it.

This is documented at https://developer.arm.com/documentation/100748/0624/Alignment-support-in-Arm-Compiler-for-Embedded-6/Aligned-and-unaligned-accesses.

We should remove these lowerings to support all armv7 processors.
In the future, we may want to add them back when we can handle
the alignment check separately.

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::storePair32):

Canonical link: https://commits.webkit.org/298097@main
